kube-proxy should infer the service IP family from the ClusterIP field

aojea opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug
/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

when dual-stack kube-proxy infers the service IP family from
the ClusterIP because, quoting Dan Winship

" IPFamily's semantics are odd and inconsistent, and its
  behavior and range of values may change in future releases, and
  so everyone is strongly discouraged from trying to interpret
  it, and should instead simply look at the service's ClusterIP
  or endpoints."

Since kube-proxy skip headless and externalname services we
can safely obtain the IPFamily from the ClusterIP field

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #86773

**Special notes for your reviewer**:

Services without ClusterIP, ClusterIP=None or ExternalName Type are skipped by kube-proxy
https://github.com/kubernetes/kubernetes/pull/91357#discussion_r431058024

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-proxy, in dual-stack mode, infers the service IP family from the ClusterIP instead of using the Service.Spec.IPFamily field
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea committed:
kube-proxy dual-stack infers IP family from ClusterIP

when dual-stack kube-proxy infers the service IP family from
the ClusterIP because ipFamily field is going to be deprecated.

Since kube-proxy skip headless and externalname services we
can safely obtain the IPFamily from the ClusterIP field

Signed-off-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

aojea commented on pkg/proxy/metaproxier/meta_proxier.go:
```diff
@@ -214,16 +245,16 @@ func endpointsIPFamily(endpoints *v1.Endpoints) (*v1.IPFamily, error) {
 
 	ipv4 := v1.IPv4Protocol
 	ipv6 := v1.IPv6Protocol
-	if utilnet.IsIPv6String(address.IP) {
```
The problem with this approach is that this does not handle the case with invalid IPs because we are assuming: "if is not IPv6 then it has to be IPv4". 
We had a similar bug with the legacy endpoint controller and headless services, where ClusterIP=None, 
don't know if this is possible with the endpoints Address.IP field, maybe it was already validated?

danwinship replied:
endpoint IPs will always be actual IPs, not `"None"`

aojea commented:
/assign @danwinship @thockin

aojea commented:
tested locally with KIND and dual-stack
```Ran 6 of 5098 Specs in 120.899 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 5092 Skipped
PASS
aojea@elder:~/go/src/k8s.io/kubernetes$ bazel-bin/test/e2e/e2e.test -context kind-kind -ginkgo.focus="DualStack" -ginkgo.skip="internal node ip" -num-nod
es 2
```

the skip is because kubelet without a cloud controller only reports one IP address, this was already discussed with sig-testing and are asking to remove that test from the e2e, but that's not relevant for this PR

aojea commented:
/retest
known flakes

danwinship commented on pkg/proxy/metaproxier/meta_proxier.go:
```diff
@@ -98,6 +112,23 @@ func (proxier *metaProxier) OnServiceSynced() {
 
+// serviceIPFamily that returns IPFamily of service or error if
+// failed to identify the IP family.
+func serviceIPFamily(service *v1.Service) (*v1.IPFamily, error) {
```
I guess this needs to look at `IPFamily` as well (to deal with headless services), and needs to fall back rather than erroring out if neither `ClusterIP` nor `IPFamily` is usable. Basically, this should be the same as `endpointutil.IsIPv6Service` in https://github.com/kubernetes/kubernetes/pull/91399/commits/ad313391 and maybe we should put that somewhere they can share it?

aojea replied:
that brings me one question, does kube-proxy need to care about headless or externalName service types?
I mean, does kube-proxy does something for a service that does not have the clusterIP set?

danwinship replied:
oh, duh, I was thinking "headless services still have endpoints so it needs to care", but that's wrong. I think you're right. You'll need to look through the code to confirm that

aojea replied:
heh, interesting exercise. check the stack trace :)

OnServiceAdd() -> OnServiceUpdate() -> proxier.serviceChanges.Update()

https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/iptables/proxier.go#L536-L548

proxier.serviceChanges.Update() -> serviceToServiceMap()

https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/service.go#L241-L262

serviceToServiceMap() -> utilproxy.ShouldSkipService()
	https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/service.go#L306-L313
	
utilproxy.ShouldSkipService() -> IsServiceIPSet()
	https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/util/utils.go#L148-L161

and  IsServiceIPSet()	
	https://github.com/kubernetes/kubernetes/blob/6a5f0e6eda5d4d736adaa2b3dd08c4366153bacf/pkg/apis/core/helper/helpers.go#L267-L271

danwinship commented:
So I forgot to bring this up in the mailing list thread, but there's another bug in the metaproxier, which is that you can change a Service's IPFamily by moving it from `ClusterIP` to `ExternalName` and back to `ClusterIP`, and the metaproxier doesn't cope with that. Basically it should treat a `ClusterIP`→`ExternalName` Update as being a Delete, and a `ExternalName`→`ClusterIP` Update as being an Add. (Right? Because the actual proxiers will just ignore `ExternalName` services anyway.)

aojea commented:
/hold

https://github.com/kubernetes/kubernetes/pull/91357#discussion_r430069812

Let me check better all possible combinations, implications, types of services, ...

danwinship commented:
/retitle kube-proxy should infer the service IP family from the ClusterIP field

danwinship commented on pkg/proxy/metaproxier/meta_proxier.go:
```diff
@@ -62,33 +64,47 @@ func (proxier *metaProxier) SyncLoop() {
-		proxier.ipv4Proxier.OnServiceAdd(service)
+	// if ClusterIP is "None" or empty, or ExternalName service skip proxying
+	svcName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+	if utilproxy.ShouldSkipService(svcName, service) {
```
Not sure the comment is needed... `ShouldSkipService` makes it pretty clear what's going on here

aojea replied:
removed

danwinship reviewed (commented):
I had said before that we should deal with the case of a service changing to/from ExternalName as well, but it looks like we don't totally deal with that correctly even in the single-proxier case, and it ought to be done in `pkg/proxy/service.go` or `pkg/proxy/config/config.go`... So I think this is right

aojea commented:
/retest

aojea commented:
/retest
known issues

aojea commented:
/retest
/hold cancel
tested locally with KIND and works

aojea commented:
/close 

there is a new plan 😄 

https://github.com/kubernetes/enhancements/pull/1679

danwinship commented:
> there is a new plan

but it doesn't change the need for this PR... in fact, the new plan involves killing off the `ipFamily` field (and replacing it with `ipFamilies`, but it will still be the case that kube-proxy ought to only be looking at `clusterIPs`, not `ipFamilies`)

aojea commented:
> > there is a new plan
> 
> but it doesn't change the need for this PR... in fact, the new plan involves killing off the `ipFamily` field (and replacing it with `ipFamilies`, but it will still be the case that kube-proxy ought to only be looking at `clusterIPs`, not `ipFamilies`)

right, but I think is better waiting until the `clusterIPs` field is available

danwinship commented:
/reopen
the plan from kubernetes/enhancements#1679 is not be implemented until 1.20, and this fixes a sometimes panic in 1.19

aojea commented:
/retest

aojea commented:
/retest

danwinship commented on pkg/proxy/metaproxier/meta_proxier.go:
```diff
@@ -62,33 +64,44 @@ func (proxier *metaProxier) SyncLoop() {
-	if *(service.Spec.IPFamily) == v1.IPv4Protocol {
-		proxier.ipv4Proxier.OnServiceAdd(service)
+	svcName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+	if utilproxy.ShouldSkipService(svcName, service) {
```
Just make `ShouldSkipService` not take the `svcName` arg; it can just use `service.Namespace` and `service.Name` if it needs to log a message (which it normally doesn't). (And none of the other existing callers have a `NamespaceName` object in advance either.) (Also, userspace/proxier.go is redundantly re-logging the same message that `ShouldSkipService` logs, which you should fix too.) (So, one commit to fix the existing `ShouldSkipService` usage, then a second commit to use it in the metaproxier.)

aojea replied:
done

aojea committed:
kube-proxy ShouldSkipService takes only one argument

instead of receiving the service name and namespace we
can obtain it from the service object directly.

Signed-off-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

danwinship commented:
/lgtm
/approve

danwinship commented:
/lgtm

anguslees commented:
Is this likely to be backported?

Currently, my k8s 1.18.4 kube-proxy immediately explodes with a null pointer deref, because the code before this PR tried to do `*(service.Spec.IPFamily)`

aojea commented:
> Is this likely to be backported?
> 
> Currently, my k8s 1.18.4 kube-proxy (with ipvs and dualstack featureflag) immediately explodes with a null pointer deref, because the code before this PR tried to do `*(service.Spec.IPFamily)`

We have many discussion about this topic in the past, my take away is that for alpha features is preferred to not do backports, however, if we find a good reason I think it will be accepted. However, with the new dual-stack API changes and moving from `IPFamily` to `IPfamilies`, do you think is worth the effort?

anguslees commented:
> for alpha features is preferred to not do backports

:+1: I agree with this principle.
